include/flang/Optimizer/Dialect/FIRType.h

```diff
@@ -0,0 +1,357 @@
+bool isa_fir_type(mlir::Type t);
+
+/// Is `t` any of the Standard dialect types?
+bool isa_std_type(mlir::Type t);
+
+/// Is `t` any of the FIR dialect or Standard dialect types?
+bool isa_fir_or_std_type(mlir::Type t);
+
+/// Is `t` a FIR dialect type that implies a memory (de)reference?
+bool isaMemref(mlir::Type t);
+
+/// Is `t` a FIR dialect aggregate type?
+bool isanAggregate(mlir::Type t);
+
+/// Extract the `Type` pointed to from a FIR memory reference type. If `t` is
+/// not a memory reference type, then returns a null `Type`.
+mlir::Type dyn_cast_ptrEleTy(mlir::Type t);
+
+/// Boilerplate mixin template
+template <typename A, unsigned Id>
+struct IntrinsicTypeMixin {
+  static constexpr bool kindof(unsigned kind) { return kind == getId(); }
+  static constexpr unsigned getId() { return Id; }
+};
+
+class CharacterType
+    : public mlir::Type::TypeBase<CharacterType, mlir::Type,
+                                  detail::CharacterTypeStorage>,
+      public IntrinsicTypeMixin<CharacterType, TypeKind::FIR_CHARACTER> {
+public:
+  using Base::Base;
+  static CharacterType get(mlir::MLIRContext *ctxt, KindTy kind);
+  KindTy getFKind() const;
+};
+
+class CplxType : public mlir::Type::TypeBase<CplxType, mlir::Type,
+                                             detail::CplxTypeStorage>,
+                 public IntrinsicTypeMixin<CplxType, TypeKind::FIR_COMPLEX> {
+public:
+  using Base::Base;
+  static CplxType get(mlir::MLIRContext *ctxt, KindTy kind);
+  KindTy getFKind() const;
+};
+
+class IntType
+    : public mlir::Type::TypeBase<IntType, mlir::Type, detail::IntTypeStorage>,
+      public IntrinsicTypeMixin<IntType, TypeKind::FIR_INT> {
+public:
+  using Base::Base;
+  static IntType get(mlir::MLIRContext *ctxt, KindTy kind);
+  KindTy getFKind() const;
+};
+
+class LogicalType
+    : public mlir::Type::TypeBase<LogicalType, mlir::Type,
+                                  detail::LogicalTypeStorage>,
+      public IntrinsicTypeMixin<LogicalType, TypeKind::FIR_LOGICAL> {
+public:
+  using Base::Base;
+  static LogicalType get(mlir::MLIRContext *ctxt, KindTy kind);
+  KindTy getFKind() const;
+};
+
+class RealType : public mlir::Type::TypeBase<RealType, mlir::Type,
+                                             detail::RealTypeStorage>,
+                 public IntrinsicTypeMixin<RealType, TypeKind::FIR_REAL> {
+public:
+  using Base::Base;
+  static RealType get(mlir::MLIRContext *ctxt, KindTy kind);
+  KindTy getFKind() const;
+};
+
+// FIR support types
+
+class BoxType
+    : public mlir::Type::TypeBase<BoxType, mlir::Type, detail::BoxTypeStorage> {
+public:
+  using Base::Base;
+  static BoxType get(mlir::Type eleTy, mlir::AffineMapAttr map = {});
+  static bool kindof(unsigned kind) { return kind == TypeKind::FIR_BOX; }
+  mlir::Type getEleTy() const;
+  mlir::AffineMapAttr getLayoutMap() const;
+
+  static mlir::LogicalResult
+  verifyConstructionInvariants(const mlir::AttributeStorage *, mlir::Type eleTy,
```

Comment:
The first param of these methods should be an `mlir::Location`.

Reply:
For some reason, a recent rebase to the tip of mono broke all these `verifyConstructionInvariants` functions. The parameter types were changed to sneak these past the compilers.

Reply:
Yeah, that was my fault. The original parameter was an `Optional<mlir::Location`, but I changed it to `mlir::Location`. A raw attribute storage is passed instead of a proper `mlir::Location` to avoid cyclic includes in the base storage uniquer class. I really need to add a static assert of some kind to ensure that clients still use Location.

Reply:
Ah. OK, let me try that...